x/oracle/abci.go

```diff
@@ -0,0 +1,155 @@
+package oracle
+
+import (
+	"time"
+
+	"github.com/umee-network/umee/x/oracle/keeper"
+	"github.com/umee-network/umee/x/oracle/types"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// IsPeriodLastBlock returns true if we are at the last block of the period
+func IsPeriodLastBlock(ctx sdk.Context, blocksPerPeriod uint64) bool {
+	return ((uint64)(ctx.BlockHeight())+1)%blocksPerPeriod == 0
+}
+
+// EndBlocker is called at the end of every block
+func EndBlocker(ctx sdk.Context, k keeper.Keeper) error {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
+	params := k.GetParams(ctx)
+	if IsPeriodLastBlock(ctx, params.VotePeriod) {
+
+		// Build claim map over all validators in active set
+		validatorClaimMap := make(map[string]types.Claim)
+
+		maxValidators := k.StakingKeeper.MaxValidators(ctx)
+		iterator := k.StakingKeeper.ValidatorsPowerStoreIterator(ctx)
+		defer iterator.Close()
+
+		powerReduction := k.StakingKeeper.PowerReduction(ctx)
+
+		i := 0
+		for ; iterator.Valid() && i < int(maxValidators); iterator.Next() {
+			validator := k.StakingKeeper.Validator(ctx, iterator.Value())
+
+			// Exclude not bonded validator
+			if validator.IsBonded() {
+				valAddr := validator.GetOperator()
+				validatorClaimMap[valAddr.String()] = types.NewClaim(validator.GetConsensusPower(powerReduction), 0, 0, valAddr)
+				i++
+			}
+		}
+
+		// These are the denoms we want votes on
+		// TODO : Get these from the x/leverage module's accepted list of tokens
+		// Ref https://github.com/umee-network/umee/issues/281
+		var voteTargets []string
+		for _, v := range params.Whitelist {
+			voteTargets = append(voteTargets, v.SymbolDenom)
+		}
+
+		// Clear all exchange rates
+		k.IterateExchangeRates(ctx, func(denom string, _ sdk.Dec) (stop bool) {
+			k.DeleteExchangeRate(ctx, denom)
+			return false
+		})
+
+		// Organize votes to ballot by denom
+		// NOTE: **Filter out inactive or jailed validators**
+		// NOTE: **Make abstain votes to have zero vote power**
+		voteMap := k.OrganizeBallotByDenom(ctx, validatorClaimMap)
+
+		// Iterate through ballots and update exchange rates; drop if not enough votes have been achieved.
+		for denom, ballot := range voteMap {
```

Comment:
Is this taken from Terra's codebase? This seems very problematic to me. We're iterating over a map and updating state which is non-deterministic because map iteration order is non-deterministic.